pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,261 @@
+		args = append(args, "--ignore-hostname-annotation")
+	}
+
+	//TODO: Add logic for the CRD source.
+
+	return &corev1.Container{
+		Name:                     name,
+		Image:                    image,
+		ImagePullPolicy:          corev1.PullIfNotPresent,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		Args:                     args,
+	}, nil
+}
+
+// createExternalDNSDeployment creates the given deployment using the reconciler's client.
+func (r *reconciler) createExternalDNSDeployment(ctx context.Context, depl *appsv1.Deployment) error {
+	if err := r.client.Create(ctx, depl); err != nil {
+		return fmt.Errorf("failed to create externalDNS deployment %s/%s: %w", depl.Namespace, depl.Name, err)
+	}
+	r.log.Info("created externalDNS deployment", "namespace", depl.Namespace, "name", depl.Name)
+	return nil
+}
+
+// updateExternalDNSDeployment updates the in-cluster externalDNS deployment.
+// Returns a boolean if an update was made, and an error when relevant.
+func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, desired *appsv1.Deployment) (bool, error) {
+	// don't always update (or do simple DeepEqual with) the operand's deployment
+	// as this may result into a "fight" between API/admission webhooks and this controller
+	// example:
+	//  - this contorller creates a deployment with the desired fields A, B, C
+	//  - API adds some default fields D, E, F (i.e. metadata, imagePullPullPolicy, dnsPolicy)
+	//  - mutating webhooks add default values to fields E, F
+	//  - this controller gets into reconcile loop and starts all from step1
+	//  - checking that fields A, B, C are the same as desired would save us the before mentioned round trips:
+	changed, updated := externalDNSDeploymentChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	if err := r.client.Update(ctx, updated); err != nil {
+		return false, fmt.Errorf("failed to update externalDNS deployment %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+	r.log.Info("updated externalDNS deployment", "namespace", desired.Namespace, "name", desired.Name)
+	return true, nil
+}
+
+// externalDNSDeploymentChange evaluates whether or not a deployment update is necessary.
+// Returns a boolean if an update is necessary, and the deployment resource to update to.
+func externalDNSDeploymentChanged(current, expected *appsv1.Deployment) (bool, *appsv1.Deployment) {
+	changed := false
+	updated := current.DeepCopy()
+
+	if len(current.Spec.Template.Spec.Containers) > 0 && len(expected.Spec.Template.Spec.Containers) > 0 {
+		if current.Spec.Template.Spec.Containers[0].Image != expected.Spec.Template.Spec.Containers[0].Image {
+			updated.Spec.Template.Spec.Containers[0].Image = expected.Spec.Template.Spec.Containers[0].Image
+			changed = true
+		}
+		if !cmp.Equal(current.Spec.Template.Spec.Containers[0].Args, expected.Spec.Template.Spec.Containers[0].Args) {
```

Comment:
I suppose if there's any chance the args are ordered differently, it might be better to compare them iteratively.

Reply:
Well, yes, theoretically you can edit or patch the deployment which would change the order of args. This would result into an update. Don't know about the purpose of such a change though but it can be done accidentally.

Reply:
Changed to the comparison of the sorted arg slices